### ISSUE TITLE
std::out_of_range

### ISSUE THREAD

**Issue by meshpv** [ChnMasterOG/KmerGO2#6]
terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::compare: __pos (which is 18446744073709551613) > this->size() (which is 6)
/var/spool/slurmd/job2965666/slurm_script: line 23: 30425 Aborted                 (core dumped)

**Comment by ChnMasterOG** [comment]
The code generates an exception here:
KmerGO2/kmergo/kmer_union.cpp
Line 278 >>> if (files[i].compare(files[i].length() - 8, 8, ".kmc_suf", 0, 8) == 0)

You can check whether there are files with a length of 5 in the kmc output folder(this->size() = 6). These files seem to cause the error.

**Comment by ChnMasterOG** [comment]
The code generates an exception here:
KmerGO2/kmergo/kmer_union.cpp
Line 278 >>> if (files[i].compare(files[i].length() - 8, 8, ".kmc_suf", 0, 8) == 0)

You can check whether there are files with a length of 5 in the kmc output folder(this->size() = 6). These files seem to cause the error.

**Comment by meshpv** [comment]
I think I see the probem
One cannot have the categorical .csv files (or any other files) in the same folder as the the .fasta files because KmerGO2 calls kmc3 on all of the files in the folder with *.*
instead of the files identified in the .csv files

On Mar 6, 2023, at 8:13 AM, ChnMasterOG ***@***.******@***.***>> wrote:

CAUTION: This email comes from an external source; the attachments and/or links may compromise our secure environment. Do not open or click on suspicious emails. Please click on the “Phish Alert” button on the top right of the Outlook dashboard to report any suspicious emails.


The code generates an exception here:
KmerGO2/kmergo/kmer_union.cpp
Line 278 >>> if (files[i].compare(files[i].length() - 8, 8, ".kmc_suf", 0, 8) == 0)

You can check whether there are files with a length of 5 in the kmc output folder(this->size() = 6). These files seem to cause the error.

**Comment by ChnMasterOG** [comment]
> I think I see the probem One cannot have the categorical .csv files (or any other files) in the same folder as the the .fasta files because KmerGO2 calls kmc3 on all of the files in the folder with *.* instead of the files identified in the .csv files On Mar 6, 2023, at 8:13 AM, ChnMasterOG ***@***.******@***.***>> wrote: CAUTION: This email comes from an external source; the attachments and/or links may compromise our secure environment. Do not open or click on suspicious emails. Please click on the “Phish Alert” button on the top right of the Outlook dashboard to report any suspicious emails. The code generates an exception here: KmerGO2/kmergo/kmer_union.cpp Line 278 >>> if (files[i].compare(files[i].length() - 8, 8, ".kmc_suf", 0, 8) == 0) You can check whether there are files with a length of 5 in the kmc output folder(this->size() = 6). These files seem to cause the error. — Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FChnMasterOG%2FKmerGO2%2Fissues%2F6%23issuecomment-1456113136&data=05%7C01%7Cmark.siddall%40einsteinmed.edu%7C2f899c9152ff470bee8b08db1e449d19%7C9c01f0fd65e040c089a82dfd51e62025%7C0%7C0%7C638137052268622797%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=5tJ59DnRxWYrIGPCtfmXgUOB%2BYVUkU2ZPKRgkL6hE1o%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FA6JE2VDGDNZKS6Z7KSALFKLW2XPIRANCNFSM6AAAAAAVQGAKYQ&data=05%7C01%7Cmark.siddall%40einsteinmed.edu%7C2f899c9152ff470bee8b08db1e449d19%7C9c01f0fd65e040c089a82dfd51e62025%7C0%7C0%7C638137052268779023%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=axA6NQiErHoWJt7%2F%2B0VlrXj0%2F4DV8TsJvmJSYJNPNfk%3D&reserved=0>. You are receiving this because you authored the thread.Message ID: ***@***.***>

You are right. You need to put the .csv file in another folder and try again.

**Comment by meshpv** [comment]
KmerGO2 is running well now. Thank you.

If you could, Please include in the GitHub README for KmerGO2, the following items that I think would be useful for the user to know:

1) KmerGO2 must be called from its own directory because otherwise it cannot find the bin/ directory
For example if I invoke the app /users/msiddall/myapps/KmerGO2/KmerGO2
from another directory, like  /users/msiddall/data/ KmerGO2 looks for /gs/gsfs0/users/msiddall/data/bin/
When it is actually
/users/msiddall/myapps/KmerGO2/bin/

2) the directory path reference by -i must have ONLY the relevant .fasta (or .fastq) files in that directory and nothing else

3) -m 0 assumes only two categories in the file denoted with -t

I do not know if this is necessary or not, but I decided to specify different directories for -o and -e just for safety sake.

Thank you again.


On Mar 7, 2023, at 11:35 AM, ChnMasterOG ***@***.******@***.***>> wrote:


You are right. You need to put the .csv file in another folder and try again.

**Comment by ChnMasterOG** [comment]
You are welcome! Thanks for your suggestions. I will add them to the README later.

**Comment by meshpv** [comment]
I am afraid I am still encountering issues.

With the original KmerGO, on a gpu, and 350GB
It refuses to do the kmer union

Step2: k-mer union
Status = -9
Parameters error? Processes break down.

With KmerGOs, on a gpu, and 350GB
I get
in1: 100%
*********
Stage 1: 100%
Stage 2: 100%
in1: 100%
**********************
Stage 1: 100%
Stage 2: 100%
in1: 100%
/var/spool/slurmd/job2973948/slurm_script: line 22:  2357 Bus error               (core dumped)

**Comment by ChnMasterOG** [comment]
It seems to be the problem of the slurm environment.
I don't know this problem, but I found a similar bug: 
https://github.com/aws/aws-parallelcluster/issues/3629
I hope it will help you.
